Fix use of HTTPS proxies #1070

jakob-keller opened the pull request:
### Description of Change
This PR fixes support for HTTPS proxies.

### Assumptions
Connection to the HTTPS proxy and to the HTTPS endpoint may use a single `ssl.SSLContext`.

### Checklist for All Submissions
* [x] I have added change info to [CHANGES.rst](https://github.com/aio-libs/aiobotocore/blob/master/CHANGES.rst)
* [x] If this is resolving an issue (needed so future developers can determine if change is still necessary and under what conditions) (can be provided via link to issue with these details): closes #1070 
  * [ ] Detailed description of issue
  * [ ] Alternative methods considered (if any)
  * [x] How issue is being resolved: Use separate aiohttp sessions per proxy URL and merge any relevant configuration options into a single `ssl.SSLContext` to be used both for connections to the HTTPS proxy as well as the HTTPS endpoint
  * [ ] How issue can be reproduced
* [ ] If this is providing a new feature  (can be provided via link to issue with these details):
  * [ ] Detailed description of new feature
  * [ ] Why needed
  * [ ] Alternatives methods considered (if any)

### Checklist when updating botocore and/or aiohttp versions

* [ ] I have read and followed [CONTRIBUTING.rst](https://github.com/aio-libs/aiobotocore/blob/master/CONTRIBUTING.rst#how-to-upgrade-botocore)
* [ ] I have updated test_patches.py where/if appropriate (also check if no changes necessary)
* [ ] I have ensured that the awscli/boto3 versions match the updated botocore version

thehesiod commented on aiobotocore/httpsession.py:
```diff
@@ -93,25 +92,42 @@ def __init__(
+            if self._cert_file:
+                context.load_cert_chain(self._cert_file, self._key_file)
+
+            # inline self._setup_ssl_cert
```
I'd prefer we kept this logic in `_setup_proxy_ssl_context` as it'll be easier to diff going forwards.  There's a flaw in this approach though because the proxy context depends on if the host we're connecting to is an IPV4 address or not: https://github.com/boto/botocore/blob/1.34.23/botocore/httpsession.py#L396-L397.  I think we need to implement a context per proxy url unfortunately.

thehesiod replied:
which means a tcp connector per proxy url I think :(

jakob-keller replied:
> which means a tcp connector per proxy url I think :(

I'm afraid, you're right. I'll try to address that in an updated PR.

thehesiod replied:
sorry :]

thehesiod commented:
sorry for delay!  Kids got hand foot mouth sequentially and I've been working on getting our condo back on the market due to tenant destruction (pain) with no money (they didn't pay).  Will get to this ASAP.

jakob-keller commented:
> sorry for delay! Kids got hand foot mouth sequentially and I've been working on getting our condo back on the market due to tenant destruction (pain) with no money (they didn't pay). Will get to this ASAP.

Ouch, I feel ya! Take your time!

thehesiod commented:
is it possible to use their proxy manager concept to make the logic more similar?  would help to maintain going forwards.

jakob-keller commented:
> is it possible to use their proxy manager concept to make the logic more similar? would help to maintain going forwards.

I don't see how. We don't use (proxy) managers at all.

We could create a `._get_session()` method that extracts relevant logic from `.send()`, if that is what you propose. Should I give it a try?

jakob-keller commented:
I just rebased the PR onto main and added a commit that pulls out session creation into `._get_session()`. It makes the code a bit more readable, but introduces another round trip through the event loop, which could have a minor performance impact. Let me know, what you think. I can drop the commit easily.

thehesiod commented:
ok looks great, we really need to add a proxy test later though.  One suggestion, replace:
```python
            chunked = None
            if headers_.get('Transfer-Encoding', '').lower() == 'chunked':
                # aiohttp wants chunking as a param, and not a header
                headers_.pop('Transfer-Encoding', '')
                chunked = True
```
with something like this function after `_get_session`
```python
    def _chunked(self, headers):
        transfer_encoding = headers.get('Transfer-Encoding', '')
        if chunked := transfer_encoding.lower() == 'chunked':
            # aiohttp wants chunking as a param, and not a header
            headers.pop('Transfer-Encoding', '')
        return chunked
```
so it matches botocore some more.

thehesiod commented:
oh and then `chunked=self._chunked(headers_),`

thehesiod commented:
thank you and sorry for the delays!  finally finished the painting, now for all the other 1000 tiny things to fix lol

jakob-keller commented:
I'm not sure we need to deal with the chunked parameter at all. To me it seems that [aithttp is able to handle that internally](https://github.com/aio-libs/aiohttp/blob/0e91eb0bd94298e501084f245df44cc1772c49e5/aiohttp/client_reqrep.py#L435) at least since https://github.com/aio-libs/aiohttp/commit/2c858346b6b82a703e9b187ee26a95f3f6772a67.

jakob-keller commented on aiobotocore/httpsession.py:
```diff
@@ -195,20 +215,15 @@ async def send(self, request):
                 request.method,
                 url=url,
-                chunked=chunked,
+                chunked=self._chunked(headers_),
```
I'm not sure we need to deal with the chunked parameter at all. To me it seems that [aithttp is able to handle that internally](https://github.com/aio-libs/aiohttp/blob/0e91eb0bd94298e501084f245df44cc1772c49e5/aiohttp/client_reqrep.py#L435) at least since https://github.com/aio-libs/aiohttp/commit/2c858346b6b82a703e9b187ee26a95f3f6772a67.

thehesiod replied:
hmm, seems like there's test coverage so it seems like you could pull it out to see what happens (assuming it's an integration test to ensure it's testing w/ aiohttp and moto)

thehesiod replied:
although I'd open an issue with comments for better testing in this area as well to do later unless you feel very motivated :)

jakob-keller replied:
> although I'd open an issue with comments for better testing in this area as well to do later unless you feel very motivated :)

Unit tests in what's still an unfamiliar codebase are not my strong suit :-/

jakob-keller replied:
> Let's see what happens: #1096

CI doesn't complain, but I'm not sure that's enough evidence. Maybe aiohttp folks can weigh in?

thehesiod replied:
we can just add a test later, i think werkzerg via moto should be able to handle it

jakob-keller commented:
I have no idea, why CI is failing. This should be unrelated to this PR.

thehesiod commented:
hmm: 
```            if report.has_errors():
>               raise ParamValidationError(report=report.generate_report())
E               botocore.exceptions.ParamValidationError: Parameter validation failed:
E               Invalid type for parameter TopicArn, value: None, type: <class 'NoneType'>, valid types: <class 'str'>
```
https://github.com/aio-libs/aiobotocore/actions/runs/8129847790/job/22217412080?pr=1077

thehesiod commented:
hmm, also seeing an xml parse error, I wonder if it's moto that got upgraded

jakob-keller commented:
> I have no idea, why CI is failing. This should be unrelated to this PR.

For some reason chunked=None is not equivalent to chunked=False. I just reverted to the old logic and now CI is passing.

thehesiod commented:
thank you so much, sorry for delay!

jakob-keller commented:
Thanks for the thorough review! Much appreciated!
